src/program/latex/mod.rs

```diff
@@ -0,0 +1,1530 @@
+    /// (String) as the key. If a wire exists with the same name, then the
+    /// contents of the new wire are added to it by updating the next column
+    /// using the Quantikz command associated with its attributes (e.g. gate,
+    /// do_nothing, etc).
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes HashMap<String, Box<Circuit>>
+    /// `wire` - the wire to be pushed or updated to in circuits
+    fn push_wire(&mut self, wire: Wire) -> Result<(), LatexGenError> {
+        let qubit = wire.name;
+
+        // find wire in circuit collection
+        match self.circuit.get_mut(&wire.name) {
+            Some(wire_in_circuit) => {
+                // get the new gate from the wire and insert into existing wire
+                if let Some(gate) = wire.gates.get(&self.column) {
+                    // add gates to wire in circuit
+                    wire_in_circuit.gates.insert(self.column, gate.to_string());
+                }
+
+                // add modifiers to gate in circuit
+                if let Some(modifier) = wire.modifiers.get(&self.column) {
+                    wire_in_circuit
+                        .modifiers
+                        .insert(self.column, modifier.to_vec());
+                }
+
+                // add modifiers to gate in circuit
+                if let Some(parameters) = wire.parameters.get(&self.column) {
+                    wire_in_circuit
+                        .parameters
+                        .insert(self.column, parameters.to_vec());
+                }
+            }
+            _ => (),
+        }
+
+        // initalize relationships between multi qubit gates
+        if let Some(wire) = self.circuit.get(&qubit) {
+            // get the newly added gate if any at the column it was added
+            if let Some(gate) = wire.gates.get(&self.column) {
+                // tag relationships for multi qubit gates
+                if gate.starts_with('C') {
+                    // add the qubits to the set of related qubits in the current column
+                    if let Some(qubits) = self.relationships.get_mut(&self.column) {
+                        // ensure relationships are valid
+                        for _self in qubits.iter() {
+                            // qubit cannot control and target itself
+                            if *_self == qubit {
+                                return Err(LatexGenError::FoundCNOTWithNoTarget);
+                            }
+                        }
+
+                        qubits.push(qubit);
+                    } else {
+                        self.relationships.insert(self.column, vec![qubit]);
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Display for Diagram {
+    /// Returns a result containing the Diagram Circuit as LaTeX string which
+    /// can be input into the body of the Document.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
Rather than build a `String` and write it at the end, just `write!()` the parts you'd be adding to the string. It is more straightforward and less expensive.

Reply:
Refactored using `write!()` as opposed to building a `String` [Use write! in Diagram fmt, reduce/reformat docs.](https://github.com/rigetti/quil-rs/pull/145/commits/6a2fd9e508c61782156ac7e9e3829320153c0df8)